history-network.md

```diff
@@ -0,0 +1,300 @@
+```
+
+A node is expected to maintain `radius` information for each node in its local node table. A node's `radius` value may fluctuate as the contents of its local key-value store change.
+
+### Protocol Messages
+
+All protocol messages are transmitted via the `TALKREQ` and `TALKRESP` messages defined by the Discovery v5 protocol.
+
+Below is the general message encoding for all `TALKREQ` messages.
+
+```
+protocol              = portal-history
+request               = protocol-message-type || protocol-message
+protocol-message-type = uint8
+protocol-message      = byte array
+```
+
+Below is the general message encoding for all `TALKRESP` messages.
+
+```
+response              = protocol-message-type || protocol-message
+protocol-message-type = uint8
+protocol-message      = byte array
+```
+
+All `protocol-message` values are encoded and decoded according to SSZ sedes.
+
+If the size of a message would exceed the max packet size of 1280 bytes, then the transmission of the message contents would occur over [uTP](https://github.com/ethereum/stateless-ethereum-specs/blob/master/discv5-utp.md).
+
+uTP connections are initiated with randomly generated connection IDs. Upon sending a message with some `connection-id`, the sender should initiate a uTP stream using `connection-id`. Upon receiving a message that contains some `connection-id`, the recipient should listen for an incoming uTP connection with `connection-id`.
+
+NOTE: The `Offer`/`Accept`/`Store` messages do not conform to the request/response paradigm of Discovery v5. We plan to propose a change to the base protocol definition to loosen the language.
+
+#### Ping (0x01)
+
+Communicate `radius` information of the sender, and request `radius` information of the recipient.
+
+```
+protocol-message-type = 0x01
+protocol-message      = Container(radius: uint256)
+```
+
+#### Pong (0x02)
+
+In response to a `Ping` message, communicate `radius` information of the sender.
+
+```
+protocol-message-type = 0x02
+protocol-message      = Container(radius: uint256)
+```
+
+#### FindNode (0x03)
+
+Request a list of ENR values for nodes whose `logdistance` from the recipient's `node-id` is equal to one of the specified values.
+
+```
+protocol-message-type = 0x03
+protocol-message      = Container(distances: List[uint8, 256])
+```
+
+Each element of `distances` **MUST** be unique. A zero value in the `distances` list represents a request for the recipient's ENR value.
+
+#### FoundNodes (0x04)
+
+In response to a `FindNode` message, communicate a list of RLP-encoded ENR values for nodes whose `logdistance` value satisfies the request.
+
+```
+protocol-message-type = 0x04
+protocol-message      = Container(total: uint8, enrs: List[Bytes, 32])
+```
+
+The recipient may know of more than 32 nodes that satisfy the request. In this case, the recipient will need to send multiple `FoundNodes` messages in response to the `FindNode` message.
```

Comment:
I think the maximum here can be 16 records since that is our K-bucket size and thus it would be non-standard for a node to actually know about more than 16 nodes.

Reply:
The k-bucket size is 16, but a single response may contain nodes from multiple k-buckets, so it's possible to have more than 16 nodes to return. The discv5 wire spec does not specify that a single response contains only nodes from a single k-bucket, or at least that is not my interpretation. The spec does recommend a result limit of 16, but not for the reason above. If we want to match the recommendation, then I have no problem there.

Reply:
your interpretation is correct.  :facepalm:   feel free to ignore this comment